src/waveform/renderers/waveformrenderbeat.cpp

```diff
@@ -49,53 +63,108 @@ void WaveformRenderBeat::draw(QPainter* painter, QPaintEvent* /*event*/) {
     QPen beatPen(m_beatColor);
     beatPen.setWidthF(std::max(1.0, scaleFactor()));
-    painter->setPen(beatPen);
+
+    QPen barPen(m_barColor);
+    barPen.setWidthF(std::max(1.0, scaleFactor()*2));
+
+    QPen phrasePen(m_phraseColor);
+    phrasePen.setWidthF(std::max(1.0, scaleFactor()*2));
 
     const Qt::Orientation orientation = m_waveformRenderer->getOrientation();
     const float rendererWidth = m_waveformRenderer->getWidth();
     const float rendererHeight = m_waveformRenderer->getHeight();
 
+    // TODO Needed?
     int beatCount = 0;
 
     while (it->hasNext()) {
         double beatPosition = it->next();
         double xBeatPoint =
                 m_waveformRenderer->transformSamplePositionInRendererWorld(beatPosition);
-
         xBeatPoint = qRound(xBeatPoint);
 
         // If we don't have enough space, double the size.
-        if (beatCount >= m_beats.size()) {
+        // TODO Needed?
+        if (beatCount++ >= m_beats.size()) {
             m_beats.resize(m_beats.size() * 2);
         }
 
+        // Calculates the beat number for the current beat
+        long beatNum = round(beatPosition / beatLength);
+
+        // Selects the right pen, if we are in phrase also paints the phrase tag
+        if(beatNum % 16 == 0 && beatNum > 0) {
+        	// Selects the font
+        	QFont font; // Uses the application default
+            font.setPointSizeF(10 * scaleFactor());
+            font.setStretch(100);
+            font.setWeight(75);
+
+            QFontMetrics metrics(font);
+
+            // Calculates the size of the box
+            QString label(QString::number(beatNum/16+1));
+            QRect wordRect = metrics.tightBoundingRect(label);
+            const int marginX = 1;
+            const int marginY = 1;
+            wordRect.moveTop(marginX + 1);
+            wordRect.moveLeft(marginY + 1);
+            wordRect.setHeight(wordRect.height() + (wordRect.height()%2));
+            wordRect.setWidth(wordRect.width() + (wordRect.width())%2);
+            //even wordrect to have an even Image >> draw the line in the middle !
+
+            int labelRectWidth = wordRect.width() + 2 * marginX + 4;
+            int labelRectHeight = wordRect.height() + 2 * marginY + 4 ;
+
+            QRectF labelRect(xBeatPoint-labelRectWidth, rendererHeight-labelRectHeight, (float)labelRectWidth, (float)labelRectHeight);
+
+            // Draw the label rect
+            QColor rectColor = m_phraseColor;
+            rectColor.setAlpha(200);
+            painter->setPen(m_phraseColor);
+            painter->setBrush(QBrush(rectColor));
+            painter->drawRoundedRect(labelRect, 2.0, 2.0);
+
+            // Draw the text
+            painter->setBrush(QBrush(QColor(0,0,0,0)));
+            painter->setFont(font);
+            painter->setPen(Qt::black);
+            painter->drawText(labelRect, Qt::AlignCenter, label);
+
+            painter->setPen(phrasePen);
+        } else if(beatNum % 4 == 0 && beatNum > 0) {
```

Comment:
This four schoold become a constant variable to set it to the correct value finally

Reply:
Done.